lib/Dialect/TritonNvidiaGPU/Transforms/MMALowering.cpp

```diff
@@ -101,6 +101,70 @@ struct TCGen5MMAScaleSharedToTmemConversion
   }
 };
 
+std::pair<SmallVector<TCGen5CommitOp>, SmallVector<Value>>
+collectCommitOpsAfter(MMAv5OpInterface mmaOp) {
+  SmallVector<TCGen5CommitOp> commitOps;
+  SmallVector<Value> commitPredicates;
+  Operation *nextOp = mmaOp->getNextNode();
+
+  while (nextOp && !isa<MMAv5OpInterface>(nextOp)) {
+    if (auto commit = dyn_cast<TCGen5CommitOp>(nextOp)) {
+      commitOps.push_back(commit);
+      commitPredicates.push_back(commit.getPred());
+    }
+    nextOp = nextOp->getNextNode();
+  }
+
+  return {commitOps, commitPredicates};
+}
+
+void moveDefiningOpsBefore(Value val, Operation *target) {
+  SmallVector<Operation *> toMove;
+  std::function<void(Value)> collectOpsToMove = [&](Value val) {
+    if (auto defOp = val.getDefiningOp()) {
+      if (defOp->getBlock() == target->getBlock() &&
+          target->isBeforeInBlock(defOp)) {
+        for (Value operand : defOp->getOperands()) {
+          collectOpsToMove(operand);
+        }
```

Comment:
so we have to worry about side effects here? We are blindly hoisting ops that we may not be allowed to move

Reply:
Now checking `isPure` condition on defining ops, and if it fails, we give up merging commit.